pkg/flavor/kubernetes/multiwriter.go

```diff
@@ -0,0 +1,673 @@
+	defer log.Tracef("<<<<< HandleMultiWriterNodePublish")
+
+	// get service with matching volume-id(i.e original claim-id)
+	serviceName := fmt.Sprintf("%s%s", serviceNamePrefix, req.VolumeId)
+	service, err := flavor.GetNFSService(serviceName)
+	if err != nil {
+		log.Errorf("unable to obtain service %s volume-id %s to publish volume", serviceName, req.VolumeId)
+		return err
+	}
+	clusterIP := service.Spec.ClusterIP
+
+	// TODO: add nfs mount support in chapi
+	targetPath := req.GetTargetPath()
+	notMnt, err := mount.New("").IsLikelyNotMountPoint(targetPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			if err := os.MkdirAll(targetPath, 0750); err != nil {
+				return nil, status.Error(codes.Internal, err.Error())
+			}
+			notMnt = true
+		} else {
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+	}
+
+	if !notMnt {
+		return &csi.NodePublishVolumeResponse{}, nil
+	}
+
+	mo := req.GetVolumeCapability().GetMount().GetMountFlags()
+	if req.GetReadonly() {
+		mo = append(mo, "ro")
+	}
+
+	source := fmt.Sprintf("%s:%s", clusterIP, defaultExportPath)
+	mounter := mount.New("")
+	err = mounter.Mount(source, targetPath, "nfs", mo)
+	if err != nil {
+		if os.IsPermission(err) {
+			return nil, status.Error(codes.PermissionDenied, err.Error())
+		}
+		if strings.Contains(err.Error(), "invalid argument") {
+			return nil, status.Error(codes.InvalidArgument, err.Error())
+		}
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	return &csi.NodePublishVolumeResponse{}, nil
+}
+
+func (flavor *Flavor) IsMultiWriterVolume(volumeID string) bool {
+	serviceName := fmt.Sprintf("%s%s", serviceNamePrefix, volumeID)
+	_, err := flavor.GetNFSService(serviceName)
+	if err != nil {
+		log.Debugf("unable to obtain service %s volume-id %s", serviceName, volumeID)
+		return false
+	}
+	return true
+}
+
+func (flavor *Flavor) GetMultiWriterClaimFromClaimUID(uid string) (*core_v1.PersistentVolumeClaim, error) {
+	log.Tracef(">>>>> GetMultiWriterClaimFromClaimUid with claim %s", uid)
+	defer log.Tracef("<<<<< GetMultiWriterClaimFromClaimUid")
+
+	claims, err := flavor.claimIndexer.ByIndex("uid", uid)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to retrieve claim by uid %s", uid)
+	} else if len(claims) == 0 {
+		return nil, fmt.Errorf("Requested pvc was not found with uid %s", uid)
+	}
+
+	log.Infof("Found the following claims: %v", claims)
+	return claims[0].(*core_v1.PersistentVolumeClaim), nil
+}
+
+func (flavor *Flavor) getPvFromName(pvName string) (*core_v1.PersistentVolume, error) {
+	log.Tracef(">>>>> getPvFromName with claim %s", pvName)
+	defer log.Tracef("<<<<< getPvFromName")
+
+	pv, err := flavor.kubeClient.CoreV1().PersistentVolumes().Get(pvName, meta_v1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	return pv, nil
+}
+
+func (flavor *Flavor) makeNFSPvc(claim *core_v1.PersistentVolumeClaim) (*core_v1.PersistentVolumeClaim, error) {
```

Comment:
The name of this function confused me at first.  Can we just rename it to cloneClaim?

Reply:
sure, will change this..